### ISSUE TITLE
Incorrect RAM usage calculation when opencost/KSM scraping interval is different from node scraping interval.

### ISSUE THREAD

**Issue by r2k1** [opencost/opencost#2297]
When the scraping interval for opencost/KSM (Kubernetes State Metrics) differs from the node scraping interval, an erroneous RAM usage calculation occurs.

**To Reproduce**
1. Set the opencost scraping interval to 3 minutes (increase x3 from default 1 minute).
2. Restart prometheus
3. Allow the system to collect data for at least 2 hours.
4. Establish a port forward to the opencost API.
5. GET http://localhost:9003/allocation/compute?aggregate=namespace,controllerKind,controller,label:app,label:team,label:pod_template_hash&step=window&window=2023-11-07T22:00:00Z,2023-11-08T23:00:00Z
(Adjust the time frame accordingly)
6. Notice that the ramByteHours figure is significantly higher than the ramByteUsageAverage. (And likely different if you compare with data before the scraping interval change)

**Which version of OpenCost are you using?**
Latest at 08 November 2023

**Additional context**

The computation of memory usage cost involves the ratio between `queryRAMUsageStr` (as defined [here](https://github.com/opencost/opencost/blob/bdb6c5df38613d845998b7f5423885cb36ae2fa1/pkg/costmodel/costmodel.go#L154C26-L154C26))) and `normalization` (as defined [here](https://github.com/opencost/opencost/blob/bdb6c5df38613d845998b7f5423885cb36ae2fa1/pkg/costmodel/costmodel.go#L252)). The objective of this query is to compute the average memory usage while accounting for periods when the pod is inactive and not reporting memory usage, in contrast to simply using an `avg_over_time` aggregation.

It can be written as: 

```
sum_over_time(container_memory_working_set_bytes) / max(count_over_time(kube_pod_container_resource_requests))
```

`max(count_over_time(kube_pod_container_resource_requests))` serves as an approximate measure of the number of scrapes within a given timeframe. However, if these metrics originate from different sources with distinct scraping intervals, it leads to inaccurate RAM usage calculations. Specifically, less frequent scrapes of opencost data result in an overestimation of memory usage, whereas more frequent scrapes lead to underestimation.

Additionally, analyzing the query opencost is likely to report inflated RAM usage if k8s cluster doesn't have a single pod that's been alive for the full duration of a requested timeframe.

**Comment by r2k1** [comment]
Some ideas how it can be solved:

### 1. Using subquery

```
sum_over_time(container_memory_working_set_bytes[1h:5m]) / scalar(count_over_time(vector(1)[1h:5m]))
```

`scalar(count_over_time(vector(1)[1h:5m]))` returns an amount of expected datapoints (60m/5m = 20 in this case)

It's hard to tell how it is going to affect the query performance.

### 2. Using container lifetime.

This will require an information about container/pod lifetime in a requested interval. How to calculate container lifetime is an open question

```
container_avg_memory_usage = avg_over_time(container_memory_working_set_bytes) * container_life_time_seconds / requested_timeframe_seconds
```

**Comment by AjayTripathy** [comment]
@r2k1 thank you so much for pointing this out. I'm guessing because we set these numbers to be the same right now that we haven't noticed this. Subquery might make sense here but we'd definitely want to run a few scaling tests to ensure. 

@nikovacevic any ideas on how we might calculate container lifetime in a window?

Meanwhile, can we make sure we document the limitation that the two scrape intervals much match?

**Comment by nikovacevic** [comment]
@AjayTripathy yes -- in fact, we already do compute container lifetime within a window in `pkg/costmodel/allocation.go` because of this sort of issue. To do so, we query for the first and final timestamps of `kube_pod_container_status_running[%s]` where `%s` is your window duration: https://github.com/opencost/opencost/blob/bdb6c5df38613d845998b7f5423885cb36ae2fa1/pkg/costmodel/allocation.go#L20

In short, if we determine that this is a priority issue to fix, then I think that @r2k1's second proposal is reasonable, and aligns quite nicely with what we already do elsewhere in the code.

**Comment by AjayTripathy** [comment]
@nikovacevic can you link a snippet to where else we do this in code by chance?

**Comment by nikovacevic** [comment]
@AjayTripathy -- are you asking for a snippet other than the one I posted, but where we do the same thing? Or am I misunderstanding?

Happy to further explicate the above example, if that's helpful.  (e.g. how the query response is processed, etc.)

**Comment by AjayTripathy** [comment]
Yeah if you could link to the section with the response processing I think that will help future implementer! Thanks.

**Comment by nikovacevic** [comment]
Sure. Here's the quick "guide":

See `buildPodMap` for where we actually execute the query
https://github.com/opencost/opencost/blob/bdb6c5df38613d845998b7f5423885cb36ae2fa1/pkg/costmodel/allocation.go#L369

That function, in helpers, grabs and processes the result
https://github.com/opencost/opencost/blob/bdb6c5df38613d845998b7f5423885cb36ae2fa1/pkg/costmodel/allocation_helpers.go#L86

Processing occurs here, in `applyPodResults`
https://github.com/opencost/opencost/blob/bdb6c5df38613d845998b7f5423885cb36ae2fa1/pkg/costmodel/allocation_helpers.go#L118

Applying the pod results, then, involves setting a `Start` and an `End` on a per-result basis
https://github.com/opencost/opencost/blob/bdb6c5df38613d845998b7f5423885cb36ae2fa1/pkg/costmodel/allocation_helpers.go#L196-L197

Once you have that, you can run _other_ queries, say for the average of CPU usage. If your average CPU usage is `2.0`, and you apply that to a pod with `(Start, End)` of duration `12h` then you can say that the pod used `24.0` CPU-Hours. Add a price for that resource -- say, `0.05` $/CPU-Hr, and you can arrive at a cost of `24.0 * 0.05 = 1.2` i.e. $1.20.
